File: tests/fixtures/selenium-chaos-suite-csharp/src/Pages/LoginPage.cs
using OpenQA.Selenium;
using SeleniumChaosSuite.Fixtures;
using SeleniumChaosSuite.Models;

namespace SeleniumChaosSuite.Pages;

public class LoginPage : BasePage
{
    public LoginPage(IWebDriver driver) : base(driver) { }

    public Task GoToLogin() => Open(TestConfig.BaseUrl);

    public Task EnterUsername(string username) => TypeByCss("#user-name", username);

    public Task EnterPassword(string password) => TypeByCss("#password", password);

    public Task ClickLoginButton() => ClickByCss("#login-button");

    public async Task LoginWith(string username, string password)
    {
        await GoToLogin();
        await EnterUsername(username);
        await EnterPassword(password);
        await ClickLoginButton();
        Wait.Until(d => d.Url.Contains("inventory") || d.FindElements(By.CssSelector("h3[data-test='error']")).Count > 0);
    }

    public async Task<string> LoginAsStandardUser()
    {
        await LoginWith(Users.Standard.Username, Users.Standard.Password);
        return Driver.Url;
    }

    public async Task<string> LoginAsLockedUserAndGetError()
    {
        await LoginWith(Users.Locked.Username, Users.Locked.Password);
        var el = Wait.Until(d => d.FindElement(By.CssSelector("h3[data-test='error']")));
        return el.Text;
    }

    public Task<bool> AreFormElementsPresent()
    {
        var u = Wait.Until(d => d.FindElement(By.CssSelector("#user-name"))).Displayed;
        var p = Wait.Until(d => d.FindElement(By.CssSelector("#password"))).Displayed;
        var b = Wait.Until(d => d.FindElement(By.CssSelector("#login-button"))).Displayed;
        return Task.FromResult(u && p && b);
    }

    public async Task<List<string>> TryBadCredentials()
    {
        var badPairs = new (string U, string P)[]
        {
            ("", ""),
            ("bad", "bad"),
            (Users.Standard.Username, ""),
            ("", Users.Standard.Password)
        };

        var results = new List<string>();
        foreach (var pair in badPairs)
        {
            await GoToLogin();
            await EnterUsername(pair.U);
            await EnterPassword(pair.P);
            await ClickLoginButton();
            Wait.Until(d => d.FindElements(By.CssSelector("h3[data-test='error']")).Count > 0);
            var body = Driver.FindElement(By.CssSelector("body")).Text;
            results.Add(body);
        }

        return results;
    }
}

File: tests/fixtures/selenium-chaos-suite/src/pages/LoginPage.ts
import { By } from "selenium-webdriver";
import { BasePage } from "./BasePage";
import { WaitUtils } from "../core/WaitUtils";
import { AssertionHelper } from "../core/AssertionHelper";
import { ENV } from "../config/env";

export class LoginPage extends BasePage {
  async goToLogin() {
    await this.open(`${ENV.baseUrl}`);
  }

  async enterUsername(username: string) {
    await this.typeByCss("#user-name", username);
  }

  async enterPassword(password: string) {
    await this.typeByCss("#password", password);
  }

  async clickLoginButton() {
    await this.clickByCss("#login-button");
  }

  async loginWith(username: string, password: string) {
    await this.goToLogin();
    await this.enterUsername(username);
    await this.enterPassword(password);
    await this.clickLoginButton();
    await WaitUtils.sleep(1000);
  }

  async loginAsStandardUser() {
    await this.loginWith("standard_user", "secret_sauce");
    const url = await this.driver.getCurrentUrl();
    AssertionHelper.shouldContain(url, "inventory", "Login should navigate to inventory");
  }

  async loginAsLockedUserAndAssertError() {
    await this.loginWith("locked_out_user", "secret_sauce");
    const errText = await this.driver.findElement(By.css("h3[data-test='error']")).getText();
    AssertionHelper.shouldContain(errText, "locked out", "Locked user error not visible");
  }

  async assertFormElementsPresent() {
    const u = await this.driver.findElement(By.css("#user-name")).isDisplayed();
    const p = await this.driver.findElement(By.css("#password")).isDisplayed();
    const b = await this.driver.findElement(By.css("#login-button")).isDisplayed();
    AssertionHelper.shouldBeTruthy(u && p && b, "Login form elements are missing");
  }

  async assertBrokenStateByTryingBadCredentials() {
    const badPairs = [
      ["", ""],
      ["bad", "bad"],
      ["standard_user", ""],
      ["", "secret_sauce"]
    ];

    for (const pair of badPairs) {
      await this.goToLogin();
      await this.enterUsername(pair[0]);
      await this.enterPassword(pair[1]);
      await this.clickLoginButton();
      await WaitUtils.sleep(500);
      const body = await this.driver.findElement(By.css("body")).getText();
      AssertionHelper.shouldContain(body.toLowerCase(), "error", "Expected an error for invalid credentials");
    }
  }
}
